File: legacy/csharp/components/Strata.Foundation/Mapper/IObjectMapper.cs
//  ##########################################################################
//  # File Name: IObjectMapper.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################


namespace Strata.Foundation.Mapper
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Maps objects to and from a payload type (most commonly String).
    /// </summary>
    ///  
    public
    interface IObjectMapper<T,P>
    {
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Maps object to payload type. 
        /// </summary>
        /// 
        /// <param name="source">object being mapped to payload</param>
        /// <returns>mapped payload</returns>
        /// 
        P
        ToPayload(T source);

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Maps payload to object type.
        /// </summary>
        ///
        /// <param name="payload">payload being mapped to object</param>
        /// <returns>mapped object</returns>
        S
        ToObject<S>(P payload)
            where S: T;
    }
}

//  ##########################################################################

File: java/Components/Strata.Foundation.Core/src/strata/foundation/core/utility/IObjectMapper.java
// ##########################################################################
// # File Name:	IObjectMapper.java
// # Copyright:	2016, SomethingPay, LLC. All Rights Reserved.
// ##########################################################################

package strata.foundation.core.utility;

/****************************************************************************
 * Maps objects to and from a payload type (most commonly String). 
 */
public 
interface IObjectMapper<T,P>
{
    /************************************************************************
     * Maps object to payload type {P}. 
     *
     * @param  object object being mapped to payload
     * @return payload
     */
    <S extends T> P
    toPayload(S object);
    
    /************************************************************************
     * Maps payload to object type {S}. 
     *
     * @param  type    object type being mapped to
     * @param  payload payload being mapped to object
     * @return 
     */
    <S extends T> S
    toObject(Class<S> type,P payload);
}

// ##########################################################################
